### ISSUE TITLE
Find if LP tightening of the logit bounds pays for itself on RSL18a

### ISSUE THREAD

**Issue by vtjeng** [vtjeng/MIPVerify.jl#270]
## Summary

On the `MNIST.RSL18a_linf0.1_authors` network, LP tightening of the logit bounds costs 2.2 to 7.1 seconds for each sample. Its only measured effect removes no work. Find if the cost gives a benefit in the main solve. If it does not, use `interval_arithmetic` for this network.

## Background

`MIPVerify.maximum` calculates an upper bound and a lower bound for each of the 9 target logits. It uses the bounds for two purposes:

1. It removes a target logit when the upper bound of that logit is not more than the largest lower bound. A removed logit needs no binary variable in the disjunction.
2. It sets the big-M constants of the disjunction. The constraint is `x_max <= x_i + (1 - a_i) * (umaxi - ls[i])`. Better bounds give smaller constants and thus a tighter relaxation.

The ReLU layer of RSL18a uses `interval_arithmetic`, because the network definition sets it. The bounds of that layer are exact, because the layer is the first affine layer over a box. Thus the logit bounds are the only bounds that an algorithm can make better.

## Measurements

Conditions: L-infinity 0.1, HiGHS, model built with `interval_arithmetic`, MNIST test samples 1 to 3.

| Sample | Split ReLUs | Model build (s) | LP tightening of 20 logit bounds (s) | LP width / interval width |
| --- | --- | --- | --- | --- |
| 1 | 321 | 11.4 (includes compilation) | 7.1 | 40% to 41% |
| 2 | 260 | 0.35 | 2.2 | 35% to 38% |
| 3 | 335 | 0.74 | 4.2 | 41% to 42% |

The LP bounds are much better than the interval bounds. The width becomes approximately 40 percent of the interval width. This is expected: the logits are affine in the outputs of about 300 split ReLUs, and interval arithmetic treats those outputs as independent.

The tightening costs 5 to 10 times the cost of building the complete model.

Purpose 1 above gets no benefit. For all 3 samples, 9 of 9 target logits stay in the disjunction with the interval bounds and also with the LP bounds. No logit is removed.

Thus only purpose 2 can give a benefit. We did not measure it.

## Question to answer

Do the smaller big-M constants make the main solve faster by more than the 2.2 to 7.1 seconds that they cost?

## How to answer it

Run the RSL18a benchmark two times on the same samples. Use `--tightening interval_arithmetic` for the first run and `--tightening lp` for the second run. Compare the number of resolved samples and the main solve time of the resolved samples.

This needs two things first:

- the RSL18a benchmark harness (#269);
- a resolve rate that is not zero. At present no sample resolves at a 120-second limit, so the main solve time cannot show a difference. See #268.

## Note about an idea that we rejected

`tight_bound` can stop early when a bound proves a cutoff value. `MIPVerify.maximum` does not use this. It calls `tight_upperbound(x)` and `tight_lowerbound(x)` with the default cutoff values of minus infinity and plus infinity. Thus no logit bound solve stops early.

We proposed to give the largest lower bound as the cutoff for the upper bound solves. The measurement above shows that this saves nothing on RSL18a. Each target logit has an upper bound that is much larger than the largest lower bound, with both bound sets. No solve would stop early.

We did not measure this on WK17a. The nightly output records the bound counts but not the bound values, so the existing data cannot answer it. Do a small test before any work on this idea.

_AI collaboration: co-produced with Claude Code (Anthropic)._

**Comment by vtjeng** [comment]
I ran a paired, serial comparison on MNIST test samples 1–20. LP logit-bound tightening certified 7/20 samples within 600 seconds; interval arithmetic certified 5/20. Interval arithmetic lost the certificates for samples 15 and 18 and gained no certificates among the 13 samples that LP tightening timed out on. These results support keeping LP tightening for RSL18a with HiGHS at this time limit.

The comparison held the formulation and solver settings fixed except for `tightening_algorithm`. Both runs used commit `c4ec5f2db82727ad3a0680ed55c8b415c6b61ad6`, Julia 1.12.6, HiGHS 1.14.0 through HiGHS.jl 1.23.0, a zero-objective feasibility formulation, an L-infinity radius of 0.1, one solver thread, `parallel=off`, and `random_seed=0`. The runs disabled feasibility jump, RINS, RENS, and the root reduced-cost heuristic. The machine had a virtualized Intel Core i7-8700K with 5 physical cores and 10 logical CPUs. Every sample ran serially with a 600-second main-solve limit. The [LP baseline setup and results](https://github.com/vtjeng/MIPVerify.jl/issues/268#issuecomment-5157175775) contain the remaining environment details.

The serial batch invoked these commands once per sample and tightening treatment:

```console
julia --project=/tmp/mipv-pgd-rsl/benchmarks /tmp/export_rsl18a_model_with_tightening.jl SAMPLE MODEL.mps TIGHTENING
julia --project=/tmp/mipv-pgd-rsl/benchmarks /tmp/run_highs_mps.jl MODEL.mps 600 SOLVER.log mip_heuristic_run_feasibility_jump=false mip_heuristic_run_rins=false mip_heuristic_run_rens=false mip_heuristic_run_root_reduced_cost=false
```

| Sample | LP tightening | Interval arithmetic |
|---:|---:|---:|
| 1 | certificate, 253.02 s | certificate, 498.41 s |
| 2 | timeout | timeout |
| 3 | timeout | timeout |
| 4 | certificate, 211.94 s | certificate, 99.69 s |
| 5 | timeout | timeout |
| 6 | certificate, 458.57 s | certificate, 208.37 s |
| 7 | timeout | timeout |
| 8 | timeout | timeout |
| 9 | timeout | timeout |
| 10 | timeout | timeout |
| 11 | certificate, 252.06 s | certificate, 182.48 s |
| 12 | timeout | timeout |
| 13 | timeout | timeout |
| 14 | certificate, 433.14 s | certificate, 262.94 s |
| 15 | certificate, 190.77 s | timeout |
| 16 | timeout | timeout |
| 17 | timeout | timeout |
| 18 | certificate, 374.18 s | timeout |
| 19 | timeout | timeout |
| 20 | timeout | timeout |

On the five samples that both treatments certified, interval arithmetic completed four faster. Its summed main-solve time was 1,251.89 seconds, compared with 1,608.73 seconds for LP tightening, a 22.2% reduction. An unchanged repeat of the exact interval-arithmetic sample-4 MPS preserved the solver work at 44 nodes and 190,019 simplex iterations but moved wall time from 99.69 to 84.62 seconds. That 16.4% difference relative to the two-run mean supplies a noise-floor estimate, so the common-five wall-time result is 22.2% ± 16.4%. The solver-work totals moved in the same direction on those five samples: interval arithmetic used 527 versus 749 nodes, a 29.6% reduction, and 1,919,111 versus 2,187,393 simplex iterations, a 12.3% reduction. The five-sample timing result remains exploratory because the noise estimate comes from one repeated MPS and the sample population is small.

The 600-second cap censors every timeout. Assigning 600 seconds to each timeout only to total the observed workload, LP tightening used 9,974.86 seconds of main-solve time and interval arithmetic used 10,252.51 seconds. A warmed, alternating-order formulation measurement over the same 20 samples found that LP tightening added 57.97 seconds in total: 2.90 seconds per sample on average, with a 2.59-second median and a 1.75–5.50-second range. Combining those formulation measurements with the capped solver totals produces 10,040.47 seconds for LP tightening and 10,260.15 seconds for interval arithmetic. LP tightening therefore reduced the observed end-to-end total by 219.68 seconds while certifying two additional samples. The censored total describes this 20-sample workload; it does not estimate uncapped solve time.

A direct sample-1 MPS diff isolates the mechanism. LP tightening changed the nine maximum-disjunction coefficients and right-hand sides plus the bound on their shared maximum variable. The nine big-M coefficients were 40.1–44.9% of their interval-arithmetic values, with a 41.4% mean. For sample 1, the row count, variable count, binary count, ReLU constraints, and MPS ordering were identical.

The earlier PGD screen found validated attacks for samples 2, 9, and 19. Both MIP treatments timed out on all three. A practical PGD-then-MIP pipeline therefore resolves 10/20 samples with LP tightening, comprising 3 attacks and 7 certificates, and 8/20 with interval arithmetic, comprising 3 attacks and 5 certificates.

My recommendation is to keep LP tightening for the RSL18a logit bounds under HiGHS. Its roughly three-second average construction cost paid for two certificates that interval arithmetic missed, and interval arithmetic did not rescue any LP timeout.

_AI collaboration: co-produced with Codex._
